### ISSUE TITLE
Unexpected token for valid JSON

### ISSUE THREAD

**Issue by jokay** [prantlf/jsonlint#15]
Given the following JSON ([launchSettings.json.zip](https://github.com/prantlf/jsonlint/files/9548330/launchSettings.json.zip)):

```json
{
  "profiles": {
    "app.service.backbone": {
      "commandName": "Project",
      "environmentVariables": {
        "DOTNET_ENVIRONMENT": "Development"
      }
    }
  }
}
```

Which outputs the following error:

```sh
# jsonlint launchSettings.json 
File: launchSettings.json
Parse error on line 1, column 1:
{  "profiles": { ...
```

Not sure why the same JSON is considered **valid** on the [demo website](https://prantlf.github.io/jsonlint/) 🧐 

![image](https://user-images.githubusercontent.com/18613935/189663233-99b34e78-e452-4222-b65a-bc1e8765d7d9.png)

Any help would be much appreciated.

**Comment by prantlf** [comment]
You had a great idea to attach the JSON file along with your report. I suspected that the first character would be something invisible, but not a whitespace. And it was!

Your file starts with [BOM for UTF-8](https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8). If you transport the file anywhere and interpret it as a text file, it will be correctly decoded as UTF-8. However, if you interpret it as JSON, you don't need BOM, because JSON is specified to be encoded in UTF-8 by default. But they are stricter in [RFC 8259, 8.1 Character Encoding](https://www.rfc-editor.org/rfc/rfc8259#section-8.1):

> Implementations MUST NOT add a byte order mark (U+FEFF) to the beginning of a networked-transmitted JSON text.

Ha! Your mistake! Hurry up and remove it from your file! :-)

But then they continue:

> In the interests of interoperability, implementations that parse JSON texts MAY ignore the presence of a byte order mark rather than treating it as an error.

Grrr! The kind implementor of a JSON-parsing library will have to handle it anyway. Why not simplifying the standard by saying just "BOM is optional"?

## Workaround

Currently, you have to strip BOM off the files that you validate.

## Fix

I'm going to add a parameter to the parser, so that the BOM can be either reported as an error (now), or ignored (in the next release).

**Comment by jokay** [comment]
Cool, thx a lot for analyzing this 💪🏼

**Comment by prantlf** [comment]
I added a command-line switch `-B, --bom` to ignore the leading BOM, similarly to `-C, --comments`, which ignores comments. Programmatically it can be enabled by setting the option `ignoreBOM` to `true`, similarly to `ignoreComments`.

When `fs.readFileSync()` from Node.js converts the result buffer to a string, it translates the leading `EFBBBF` (UTF-8 BOM) to `FEFF` (UTF-16 BOM). If this apparently invisible character is not removed, it will remain in the string and passed along.

The price for ignoring the BOM is the same as for other JSON5 features, like comments or single quotes - it will be always parsed by the custom parser and not by the built-in `JSON.parse`. (If no non-standard features are enabled, `JSON.parse` is preferred internaly for much vetter parsing speed.)

Implemented by [311c6df](https://github.com/prantlf/jsonlint/commit/311c6df75963a5b6da3984ba85541b800d751939) and released in [11.7.0](https://github.com/prantlf/jsonlint/releases/tag/v11.7.0).
